rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,309 @@
+
+    def __str__(self) -> Text:
+        # Describe where the conflict occurs in the stories
+        last_event_type, last_event_name = _get_previous_event(self._sliced_states[-1])
+        if last_event_type:
+            conflict_message = f"Story structure conflict after {last_event_type} '{last_event_name}':\n"
+        else:
+            conflict_message = (
+                f"Story structure conflict at the beginning of stories:\n"
+            )
+
+        # List which stories are in conflict with one another
+        for action, stories in self._conflicting_actions.items():
+            conflict_message += (
+                f"  {self._summarize_conflicting_actions(action, stories)}"
+            )
+
+        return conflict_message
+
+    @staticmethod
+    def _summarize_conflicting_actions(action: Text, stories: List[Text]) -> Text:
+        """Gives a summarized textual description of where one action occurs.
+
+        Args:
+            action: The name of the action.
+            stories: The stories in which the action occurs.
+
+        Returns:
+            A textural summary.
+        """
+        if len(stories) > 3:
+            # Four or more stories are present
+            conflict_description = (
+                f"'{stories[0]}', '{stories[1]}', and {len(stories) - 2} other trackers"
+            )
+        elif len(stories) == 3:
+            conflict_description = f"'{stories[0]}', '{stories[1]}', and '{stories[2]}'"
+        elif len(stories) == 2:
+            conflict_description = f"'{stories[0]}' and '{stories[1]}'"
+        elif len(stories) == 1:
+            conflict_description = f"'{stories[0]}'"
+        else:
+            raise ValueError(
+                "An internal error occurred while trying to summarise a conflict without stories. "
+                "Please file a bug report at https://github.com/RasaHQ/rasa."
+            )
+
+        return f"{action} predicted in {conflict_description}\n"
+
+
+class TrackerEventStateTuple(NamedTuple):
+    """Holds a tracker, an event, and sliced states associated with those."""
+
+    tracker: TrackerWithCachedStates
+    event: Event
+    sliced_states: List[Dict[Text, float]]
+
+    @property
+    def sliced_states_hash(self) -> int:
+        return hash(str(list(self.sliced_states)))
+
+
+def find_story_conflicts(
+    trackers: List[TrackerWithCachedStates],
+    domain: Domain,
+    max_history: Optional[int] = None,
+) -> List[StoryConflict]:
+    """Generates `StoryConflict` objects, describing conflicts in the given trackers.
+
+    Args:
+        trackers: Trackers in which to search for conflicts.
+        domain: The domain.
+        max_history: The maximum history length to be taken into account.
+
+    Returns:
+        StoryConflict objects.
+    """
+    # Use the length of the longest story for `max_history` if not specified otherwise
+    if not max_history:
+        max_history = max([len(tracker.past_states(domain)) for tracker in trackers])
```

Comment:
let's put that in a helper function `_get_length_of_longest_story()` and then we can also dump the comment (and test this in case it isn't yet)